core/server/src/main/java/alluxio/master/file/FileSystemMaster.java

```diff
@@ -1830,65 +1857,114 @@ private long renameAndJournal(LockedInodePath srcInodePath, LockedInodePath dstI
+          if (!ufs.mkdirs(ufsDirAndPerm.getFirst(), ufsDirAndPerm.getSecond())) {
+            throw new IOException(
+                ExceptionMessage.FAILED_UFS_CREATE.getMessage(ufsDirAndPerm.getFirst()));
+          }
+        }
+        boolean success;
+        if (srcInode.isFile()) {
+          success = ufs.renameFile(ufsSrcPath, ufsDstUri);
+        } else {
+          success = ufs.renameDirectory(ufsSrcPath, ufsDstUri);
+        }
+        if (!success) {
           throw new IOException(
-              ExceptionMessage.FAILED_UFS_CREATE.getMessage(ufsDirAndPerm.getFirst()));
+              ExceptionMessage.FAILED_UFS_RENAME.getMessage(ufsSrcPath, ufsDstUri));
         }
       }
-      boolean success;
-      if (srcInode.isFile()) {
-        success = ufs.renameFile(ufsSrcPath, ufsDstUri);
-      } else {
-        success = ufs.renameDirectory(ufsSrcPath, ufsDstUri);
-      }
-      if (!success) {
-        throw new IOException(
-            ExceptionMessage.FAILED_UFS_RENAME.getMessage(ufsSrcPath, ufsDstUri));
+    } catch (Exception e) {
+      // On failure, revert changes and throw exception.
+      if (!dstParentInode.removeChild(dstName)) {
+        LOG.error("Failed to revert rename changes. Alluxio metadata may be inconsistent.");
       }
+      srcInode.setName(srcName);
+      srcInode.setParentId(srcParentInode.getId());
+      throw e;
     }
 
     // TODO(jiri): A crash between now and the time the rename operation is journaled will result in
     // an inconsistency between Alluxio and UFS.
-    InodeDirectory dstParentInode = dstInodePath.getParentInodeDirectory();
-    srcParentInode.removeChild(srcInode);
-    srcParentInode.setLastModificationTimeMs(opTimeMs);
-    srcInode.setParentId(dstParentInode.getId());
-    srcInode.setName(dstPath.getName());
-    dstParentInode.addChild(srcInode);
+
+    // 4. Remove the source inode (reverting the name) from the source parent. The name must be
+    // reverted or removeChild will not be able to find the appropriate child entry since it is
+    // keyed on the original name.
+    srcInode.setName(srcName);
+    if (!srcParentInode.removeChild(srcInode)) {
+      // This should never happen.
+      LOG.error("Failed to rename within Alluxio. Alluxio and under storage may be inconsistent.");
+      srcInode.setName(dstName);
+      if (!dstParentInode.removeChild(dstName)) {
+        LOG.error("Failed to revert rename changes. Alluxio metadata may be inconsistent.");
+      }
+      srcInode.setName(srcName);
+      srcInode.setParentId(srcParentInode.getId());
+      throw new IOException("Failed to remove source path " + srcPath + " from parent");
+    }
+    srcInode.setName(dstName);
+
+    // 5. Set the last modification times for both source and destination parent inodes.
     dstParentInode.setLastModificationTimeMs(opTimeMs);
+    srcParentInode.setLastModificationTimeMs(opTimeMs);
```

Comment:
since we don't have WRITE lock on dst and src parent inode, I noticed that `setLastModificationTimeMs` now becomes thread-safe which is not documented. Could you update both the comment here (saying this requires thread-safe on parerent inodes) and the java doc on `setLastModificationTimeMs`